### DISCUSSION TITLE
High memory usage when compressing data blocks with up to 2MiB sizes with zstd.Encoder.EncodeAll()

### DISCUSSION THREAD

**Discussion by valyala** [klauspost/compress#1025] [Q&A]
It has been appeared that every [`zstd.Encoder`](https://pkg.go.dev/github.com/klauspost/compress/zstd#Encoder) instance occupies around 500MB of RAM when it is used for compressing data blocks with up to 2MiB sizes each by concurrent goroutines calling [`zstd.Encoder.EncodeAll()`](https://pkg.go.dev/github.com/klauspost/compress/zstd#Encoder.EncodeAll) - see [this comment](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7503#issuecomment-2500088591) for details.

IMHO, 500MB of RAM per every `zstd.Encoder` instance in this case sounds too much, and its memory usage can be reduced somehow.

**Comment by klauspost** [comment]
Higher levels take more memory - not only by bigger default window size.

Tips:

A) Use less instances. You seem to do that somewhat, but AFAICT you are indexing by original level, not the level used internally. Use the returned value from `idx := zstd.EncoderLevelFromZstd(compressionLevel)` to get your index.
B) Limit [Window Size](https://pkg.go.dev/github.com/klauspost/compress/zstd#WithWindowSize).
C) Use [WithLowerEncoderMem](https://pkg.go.dev/github.com/klauspost/compress/zstd#WithLowerEncoderMem)
D) Limit [concurrency](https://pkg.go.dev/github.com/klauspost/compress/zstd#WithEncoderConcurrency) - maybe on 32 bit platforms.
